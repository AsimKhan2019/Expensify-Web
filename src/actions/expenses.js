import database from '../firebase/firebase'

export const addExpense = (expenseData = {}) => {
  return (dispatch) => {

    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = {
      description,
      note,
      amount,
      createdAt
    };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch({
        type: 'ADD_EXPENSE',
        expense: {
          id: ref.key,
          ...expense
        }
      });
    });

  };
};

export const removeExpense = (id) => {
  return (dispatch) => {

    return database.ref(`expenses/${id}`).remove().then( () => {
      dispatch({
        type: 'REMOVE_EXPENSE',
        id
      });
    });
  };
};

export const editExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then( () => {
      dispatch({
        type: 'EDIT_EXPENSE',
        id,
        updates
      })
    });
  };
};

export const setExpenses = () => {
  return (dispatch) => {

    return database.ref('expenses').once('value').then( (snapshot) => {
      //Parsing
      const expenses = [];
      snapshot.forEach( (childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch({
        type: 'SET_EXPENSES',
        expenses: expenses
      });
    });
  }
};
