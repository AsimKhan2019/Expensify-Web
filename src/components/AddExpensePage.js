import React from 'react';
import ReactDom from 'react-dom';

const AddExpensePage = () => (
  <div>
    This is from my add expense page component
  </div>
);

export default AddExpensePage;
