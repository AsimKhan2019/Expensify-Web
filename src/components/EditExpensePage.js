import React from 'react';
import ReactDom from 'react-dom';

const EditExpensePage = () => (
  <div>
    This is from my edit expense page component
  </div>
);

export default EditExpensePage;
