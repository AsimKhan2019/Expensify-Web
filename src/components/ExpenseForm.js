import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
  state={
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    calendarFocused: false
  }
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState( () => ({description}));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState( () => ({note}));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      this.setState( () => ({amount}));
    }
  }

  onDateChange = (createdAt) => {
    this.setState( () => ({createdAt}));
  }

  onCalendarFocusedChange = ({focused}) => {
    this.setState( () => ({calendarFocused: focused}));
  }

  render(){
    return(
      <div>
        <form>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}

          />
          <input
            type="text"
            placeholder="Amount"
            autoFocus
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt} // momentPropTypes.momentObj or null
            onDateChange={this.onDateChange} // PropTypes.func.isRequired
            focused={this.state.calendarFocused} // PropTypes.bool
            onFocusChange={this.onCalendarFocusedChange} // PropTypes.func.isRequired
            numberOfMonths={1}
            isOutsideRange={() => false}
          />


          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >

          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
