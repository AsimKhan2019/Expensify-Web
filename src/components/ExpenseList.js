import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = ({expenses, dictionary}) => (
  <div className="content-container">
    <div className="list-header">
      <div >{dictionary.tableExpense}</div>
      <div >{dictionary.tableAmount}</div>
    </div>
    <div className="list-body">
      {
        expenses.length == 0 ? (
          <div className="list-item-message">
            <span>{dictionary.noExpenseMessage}</span>
          </div>
        ) : (
          expenses.map( (expense) => {
            return <ExpenseListItem key={expense.id} dictonary={dictionary.categories} {...expense} />
          })
        )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses.present, state.filters),
    dictionary: state.lang.dictionary
  };
};

export default connect(mapStateToProps)(ExpenseList);
