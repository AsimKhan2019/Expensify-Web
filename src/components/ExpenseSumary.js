import React from 'react';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

import replaceAll from '../utils/replaceAll';


export const ExpenseSumary = ({expensesTotal, expenseCount, hideExpenses, dictionary}) => (
  <div className="page-header">
    <div className="content-container">
      <h1
        className="page-header__title"
        dangerouslySetInnerHTML={{
        __html: replaceAll(dictionary.summaryMessageTitle, {
            "{p1}": `<span>${expenseCount}</span>`,
            "{p2}": expenseCount != 1 ? 's' : '',
            "{p3}": `<span>${numeral(expensesTotal / 100).format('$0,0.00')}</span>`
      })}}>
      </h1>
      {
        hideExpenses > 1 &&
        <h4 className="page-header__subtitle"
            dangerouslySetInnerHTML={{
              __html: replaceAll(dictionary.summaryMessageSubitle, {
                "{p1}": `<span>${hideExpenses}</span>`,
                "{p2}": hideExpenses != 1 ? 's' : ''
              })
            }}></h4>
      }
      <div className="page-header__actions">
        <Link className="button" to="/create">{dictionary.addExpenseButton}</Link>
      </div>
    </div>
  </div>
);

const mapStateToProps = (state) => {
  const expenses = selectExpenses(state.expenses.present, state.filters);
  return {
    expensesTotal: getExpensesTotal(expenses),
    expenseCount: expenses.length,
    hideExpenses: state.expenses.present.length - expenses.length,
    dictionary: state.lang.dictionary
  };
};

export default connect(mapStateToProps)(ExpenseSumary);
