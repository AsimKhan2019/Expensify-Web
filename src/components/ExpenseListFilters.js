import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate, setCategoryFilter, setPaymentMethodFilter } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

import categories from "../config/categories";
import payment_methods from "../config/payment_methods";


export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  onSelectChange = (e) => {
    e.target.value == 'date' ? this.props.sortByDate() : this.props.sortByAmount();
  }
  onCategoryChange = (e) => {
    const v = e.target.value;
    this.props.setCategoryFilter(v);
  }
  onPaymentMethodChange = (e) => {
    const v = e.target.value;
    this.props.setPaymentMethodFilter(v);
  }

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  }

  render() {
    const { dictionary } = this.props;
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input placeholder={dictionary.textFilter} className="text-input" type="text" value={this.props.filters.text} onChange={this.onTextChange} />
          </div>
          <div className="input-group__item">
            <select className="select" value={this.props.filters.sortBy} onChange={this.onSelectChange}>
              <option value="date" >{dictionary.sortByDate}</option>
              <option value="amount" >{dictionary.sortByAmount}</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
              startDatePlaceholderText={dictionary.startDatePicker}
              endDatePlaceholderText={dictionary.endDatePicker}
            />
          </div>
        </div>

        <div className="input-group">
          <div className="input-group__item input-group__fluid ">
            <select className="select" value={this.props.filters.category} onChange={this.onCategoryChange}>
              <option value="disabled" >{dictionary.filterByCategoryPlaceholder}</option>
              {categories.map(pm => <option value={pm} key={pm}>
                {dictionary.categories[pm]}
              </option>)}
            </select>
          </div>
          <div className="input-group__item input-group__fluid">
            <select className="select" value={this.props.filters.payment_method} onChange={this.onPaymentMethodChange}>
              <option value="disabled" >{dictionary.filterByPaymentMethodPlaceholder}</option>
              {payment_methods.map(pm => <option value={pm} key={pm}>
                {dictionary.payment_methods[pm]}
              </option>)}
            </select>
          </div>
        </div>
      </div>
    );
  }

}

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setCategoryFilter: (endDate) => dispatch(setCategoryFilter(endDate)),
  setPaymentMethodFilter: (endDate) => dispatch(setPaymentMethodFilter(endDate)),
})


const mapStateToProps = (state) => ({
  filters: state.filters,
  dictionary: state.lang.dictionary
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
