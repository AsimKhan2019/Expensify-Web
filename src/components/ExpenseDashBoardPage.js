import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSumary from './ExpenseSumary';
import UndoRedo from './UndoRedo';

const ExpenseDashBoardPage = () => (
  <div>
    <ExpenseSumary/>
    <ExpenseListFilters/>
    <ExpenseList/>
    <UndoRedo/>
  </div>
);

export default ExpenseDashBoardPage;
