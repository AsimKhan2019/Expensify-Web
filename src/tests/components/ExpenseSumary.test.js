import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseSumary} from '../../components/ExpenseSumary';
import expenses from '../fixtures/expenses';
import en from '../../locale/en';

test('should render ExpenseSumary with 1 expense', () => {
  const wrapper = shallow(
    <ExpenseSumary
      expenseCount={1}
      expensesTotal={5640}
      hideExpenses={0}
      dictionary={en}
    />);
  expect(wrapper).toMatchSnapshot();

});

test('should render ExpenseSumary with 2 expenses', () => {
  const wrapper = shallow(
    <ExpenseSumary
      expenseCount={2}
      expensesTotal={7895}
      hideExpenses={0}
      dictionary={en}
    />);
  expect(wrapper).toMatchSnapshot();

});
