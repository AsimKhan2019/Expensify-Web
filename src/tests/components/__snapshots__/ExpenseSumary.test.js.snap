// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSumary with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Viewing <span>1</span> expense totalling <span>$56.40</span>",
        }
      }
    />
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseSumary with 2 expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Viewing <span>2</span> expenses totalling <span>$78.95</span>",
        }
      }
    />
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
